### DISCUSSION TITLE
Troubles with pydantic models and VSCode and pyright in neovim

### DISCUSSION THREAD

**Discussion by strayer** [piccolo-orm/piccolo#989] [Q&A] [answered]
Hey,

I'm currently testing Piccolo in a side-project and ran into issues when trying to use my IDE-like neovim features with the pyright language server in combination with the tables.

My project is split into two parts, a backend with a database and a CLI that runs on other systems interacting with the backend through REST APIs. Ideally I'd like to have the tables and models in shared code so I can just create pydantic models in the CLI, serialize them and pass them on to the backend which uses the same tables and models.

Right now pyright doesn't seem to understand the models created with `piccolo.utils.pydantic.create_pydantic_model`:

<img width="824" alt="75345" src="https://github.com/piccolo-orm/piccolo/assets/310624/0521aa82-1787-4074-bf66-bee43e71c59a">
<img width="418" alt="44196" src="https://github.com/piccolo-orm/piccolo/assets/310624/79d8b521-b07d-402c-803d-f74e6c8fe099">

It seems to think that DishModel is of type BaseModel:

<img width="697" alt="32914" src="https://github.com/piccolo-orm/piccolo/assets/310624/ae6342f9-1be9-49f0-a1ed-6183dffb7c6e">

I tried the same with the `class DishModel(create_pydantic_model(...)): pass`. In that case the type is detected as DishModel, but still doesn't now anything about the fields and especially complains about the constructor:

<img width="532" alt="30047" src="https://github.com/piccolo-orm/piccolo/assets/310624/398e3307-36f9-4a2e-bbe3-c885bd3b071c">

The code itself is working fine, required fields are checked and cause an exception. But I really like to lean on correct completions and linting. The same is happening in Visual Studio Code with the Python extension.

Any ideas what I might be doing wrong? Based on other discussions here in the issues I assume this should work…

**Comment by dantownsend** [comment, marked as answer]
Yeah, there are limitations when we create Pydantic models dynamically.

In my own projects, I tend to use a mix of `create_pydantic_model` and manually created Pydantic models, depending on the use case.

I've explored added better typing support for `create_pydantic_model`, but haven't found a perfect solution.

One thing we could do instead is have a command - something like `piccolo models generate` which just outputs a `models.py` file containing the Pydantic models.

**Reply by strayer** [reply]
Ahh, I didn't really think about that. My tables will be rather simple, so I guess I'll just implement the pydantic models for the CLI myself. Maybe I can get some tests going that at least validate that the generated models equal the manually implemented ones.

**Reply by strayer** [reply]
I'll create them manually for now. If I ever get into testing those I'll update this discussion.

```python
import decimal
from piccolo.columns import Decimal, Integer, Varchar
from piccolo.table import Table
from pydantic import BaseModel


class Dish(Table):
    remote_id = Integer(required=True, unique=True)
    name = Varchar(required=True)
    original_name = Varchar(required=True)
    price = Decimal(required=True)

class DishModel(BaseModel):
    remote_id: int
    name: str
    original_name: str
    price: decimal.Decimal
```

**Reply by dantownsend** [reply]
They look right to me.

**Reply by strayer** [reply]
I might have gone a bit overboard, but I'm now generating the pydantic models instead of writing them by hand and implemented a simple test that keeps me from forgetting to keep them up-to-date:

```python
import os
import json
from pathlib import Path

import typer
from piccolo.utils.pydantic import create_pydantic_model

from ..shared.tables import Dish

cli_app = typer.Typer()


@cli_app.command()
def generate_pydantic_models():
    from datamodel_code_generator import DataModelType, InputFileType, generate

    output = Path(os.path.dirname(__file__)) / ".." / "shared" / "models.py"

    json_schema = json.dumps(
        create_pydantic_model(Dish, model_name="DishModel").model_json_schema()
    )

    generate(
        json_schema,
        input_file_type=InputFileType.JsonSchema,
        input_filename="example.json",
        output=output,
        # set up the output model types
        output_model_type=DataModelType.PydanticV2BaseModel,
    )
```

```python
from typing import Any
from piccolo.utils.pydantic import create_pydantic_model

from ...shared.tables import Dish
from ...shared.models import DishModel


def remove_keys_named_extra(schema: dict[str, Any]) -> dict[str, Any]:
    filtered_schema = {key: value for key, value in schema.items() if key != "extra"}
    filtered_schema["properties"] = {
        property_key: {
            key: value for key, value in property_value.items() if key != "extra"
        }
        for property_key, property_value in schema["properties"].items()
    }

    return filtered_schema


class TestTableModels:
    def test_dish(self):
        handmade_model = DishModel.model_json_schema()
        generated_model = create_pydantic_model(
            Dish, model_name="DishModel"
        ).model_json_schema()

        assert handmade_model == remove_keys_named_extra(generated_model)
```

Needs some more work done to support multiple tables instead of just one, but step by step.

**Reply by dantownsend** [reply]
That's cool - nice work!

**Reply by nVitius** [reply]
@strayer would you consider making a PR to add your generator to Piccolo? I think this functionality would be really useful to have.

**Reply by strayer** [reply]
> @strayer would you consider making a PR to add your generator to Piccolo? I think this functionality would be really useful to have.

Hey! Sorry, to be completely honest, I don't use Piccolo anymore. Anyone interested in building such a thing is free to use my above code snippets though.
